File: chapters/02-generative-ai-fundamentals/python/model_boundary_example.py
from dataclasses import dataclass


@dataclass(frozen=True)
class GenerationRequest:
    system_instruction: str
    user_input: str
    model: str
    temperature: float = 0.2
    max_output_tokens: int = 500


@dataclass(frozen=True)
class GenerationResponse:
    text: str
    model: str
    input_tokens: int
    output_tokens: int


class AiTextGenerator:
    def generate(self, request: GenerationRequest) -> GenerationResponse:
        raise NotImplementedError


class CustomerSupportService:
    def __init__(self, generator: AiTextGenerator) -> None:
        self._generator = generator

    def draft_reply(self, customer_message: str) -> str:
        request = GenerationRequest(
            system_instruction="Answer using only approved support guidance.",
            user_input=customer_message,
            model="configured-by-platform",
        )
        return self._generator.generate(request).text

File: chapters/02-generative-ai-fundamentals/java/ModelBoundaryExample.java
package chapter02;

/**
 * Minimal example of keeping business code independent of a concrete model provider.
 * Production implementations should add timeouts, telemetry, validation and error mapping.
 */
public final class ModelBoundaryExample {

    public interface AiTextGenerator {
        AiResponse generate(AiRequest request);
    }

    public record AiRequest(
            String systemInstruction,
            String userInput,
            String model,
            double temperature) {
    }

    public record AiResponse(
            String text,
            String model,
            int inputTokens,
            int outputTokens) {
    }

    public static final class CustomerSupportService {
        private final AiTextGenerator generator;

        public CustomerSupportService(AiTextGenerator generator) {
            this.generator = generator;
        }

        public String draftReply(String customerMessage) {
            var request = new AiRequest(
                    "Answer using only approved support guidance.",
                    customerMessage,
                    "configured-by-platform",
                    0.2);

            return generator.generate(request).text();
        }
    }
}
